staker/bold_staker.go

```diff
@@ -0,0 +1,378 @@
+	f.Uint64(prefix+".big-step-leaf-height", DefaultBoldConfig.BigStepLeafHeight, "big challenge leaf height")
+	f.Uint64(prefix+".small-step-leaf-height", DefaultBoldConfig.SmallStepLeafHeight, "small challenge leaf height")
+	f.Uint64(prefix+".num-big-steps", DefaultBoldConfig.NumBigSteps, "num big steps")
+	f.String(prefix+".validator-name", DefaultBoldConfig.ValidatorName, "name identifier for cosmetic purposes")
+	f.String(prefix+".machine-leaves-cache-path", DefaultBoldConfig.MachineLeavesCachePath, "path to machine cache")
+	f.Uint64(prefix+".assertion-posting-interval-seconds", DefaultBoldConfig.AssertionPostingIntervalSeconds, "assertion posting interval")
+	f.Uint64(prefix+".assertion-scanning-interval-seconds", DefaultBoldConfig.AssertionScanningIntervalSeconds, "scan assertion interval")
+	f.Uint64(prefix+".assertion-confirming-interval-seconds", DefaultBoldConfig.AssertionConfirmingIntervalSeconds, "confirm assertion interval")
+	f.Uint64(prefix+".check-staker-switch-interval-seconds", DefaultBoldConfig.CheckStakerSwitchIntervalSeconds, "how often to check if staker can switch to bold")
+	f.Bool(prefix+".api", DefaultBoldConfig.API, "enable api")
+	f.String(prefix+".api-host", DefaultBoldConfig.APIHost, "bold api host")
+	f.Uint16(prefix+".api-port", DefaultBoldConfig.APIPort, "bold api port")
+	f.String(prefix+".api-db-path", DefaultBoldConfig.APIDBPath, "bold api db path")
+	f.StringSlice(prefix+".track-challenge-parent-assertion-hashes", DefaultBoldConfig.TrackChallengeParentAssertionHashes, "only track challenges/edges with these parent assertion hashes")
+}
+
+type BOLDStaker struct {
+	stopwaiter.StopWaiter
+	chalManager        *challengemanager.Manager
+	blockValidator     *BlockValidator
+	rollupAddress      common.Address
+	client             bind.ContractBackend
+	lastWasmModuleRoot common.Hash
+	callOpts           bind.CallOpts
+	validatorConfig    L1ValidatorConfig
+	wallet             ValidatorWalletInterface
+}
+
+func newBOLDStaker(
+	ctx context.Context,
+	validatorConfig L1ValidatorConfig,
+	rollupAddress common.Address,
+	callOpts bind.CallOpts,
+	txOpts *bind.TransactOpts,
+	client arbutil.L1Interface,
+	blockValidator *BlockValidator,
+	statelessBlockValidator *StatelessBlockValidator,
+	config *BoldConfig,
+	dataPoster *dataposter.DataPoster,
+	wallet ValidatorWalletInterface,
+) (*BOLDStaker, error) {
+	manager, err := newBOLDChallengeManager(ctx, rollupAddress, txOpts, client, blockValidator, statelessBlockValidator, config, dataPoster)
+	if err != nil {
+		return nil, err
+	}
+	return &BOLDStaker{
+		chalManager:     manager,
+		blockValidator:  blockValidator,
+		rollupAddress:   rollupAddress,
+		client:          client,
+		callOpts:        callOpts,
+		validatorConfig: validatorConfig,
+		wallet:          wallet,
+	}, nil
+}
+
+func (b *BOLDStaker) Initialize(ctx context.Context) error {
```

Comment:
This function calls b.blockValidator.InitAssumeValid, whih cannot be called after Start (and multi_protocol does call this function after start).

I think it's o.k. to never call BOLD's Initialize if BOLD was not there before we started the node (see comment later). This is a little weird and off the normal API, and if we go that way it should be at least noted with a comment. And possibly - rename Initialize to something else so we'll know it behaves differently.

Reply:
Added a check to only call InitAssumeValid if blockValidator is not started. 
Also added a comment for Initialize stating the same.
Let me know if something else should also be done here.